libs/x-content/src/main/java/org/elasticsearch/xcontent/MemorizingXContentParser.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.xcontent;
+
+import java.io.IOException;
+
+/**
+ * A parser that parses a document preserving fields that were parsed so far in a {@link XContentBuilder}.
```

Comment:
s/preserving/copying/ to be more explicit.

I'd also note that this has performance overhead due to copying.